### DISCUSSION TITLE
Go to type declaration from hover

### DISCUSSION THREAD

**Discussion by susliko** [scalameta/nvim-metals#706] [General]
Use case: detecting the type of an expression and jumping to its declaration.

Ideally, I'd like to hover over the expression (by typing `K`), then move my cursor into the hover window (by typing `K` again), and use `gd` to navigate to any type listed in the hover.

<img src="https://github.com/user-attachments/assets/ad4ee493-131a-4ec4-81bd-d6cbdc67ad04" width="400" />

I'm curious if there's anything sensible we could do to address this use case or if there have been any attempts to make it work.

One idea I have is to invoke the `lsp_workspace_symbols` command when using `gd` inside the hover. It would be better than nothing, but still not perfect as it would open a choice window. It would be great if it could use semanticdb info somehow.

**Comment by ckipp01** [comment]
So maybe this won't be what you want all the time, but in your example if the intent is to jump to `A` instead of jumping into the hover and then attempting to navigate to a instead you can just execute a `vim.lsp.buf.type_definition` on `aaa` and it will bring you to `A`. Is that what you're attempting to do?

**Reply by susliko** [reply]
I have completely forgotten about `vim.lsp.buf.type_definition`, thanks! It certainly covers simple cases, but it seems that I have oversimplified an example. 

Oftentimes while working with expressions which type itself is a type constructor (e.g. IO/Either/Option), it would be useful to jump to definitions of its type parameters

<img src="https://github.com/user-attachments/assets/6d3a52fb-866e-44d4-86fe-f26da1cc5b66" width="400" />

in this example `type_definition` would jump to `cats.effect.IO`

**Reply by ckipp01** [reply]
I wonder if in cases like this where there is a _compound_ or _wrapped_ type (not sure what to call it) if Metals could give you the option of which to jump to. In this case if you hit `type_defintion` on here it would show you an option to navitate to `IO` or to `A`. That would be need and useful. I'd recommend creating a feature request for this in Metals. I'm not 100% sure how to achieve this without ugly hacks, but having Metals just natively support this would be great.
